### ISSUE TITLE
Multi-Container: Replace HTTP with gRPC

### ISSUE THREAD

**Issue by eddyashton** [microsoft/CCF#4160]
I believe at the moment it will be simpler to implement a working multi-container prototype using the existing HTTP transport, with the entire system as a CCF app. Eventually, we want to replace that with gRPC transport, and benefit from more flexible (async) execution. This may be integrated earlier than I expect, but I'm currently treating this as a single porting step once some demo implementation exists.

**Comment by eddyashton** [comment]
Thought about this a little more, and realised it may be a relatively small change.

We're planning for the initial demo to be HTTP/1.1, with a JSON payload. When we say "replace with gRPC", that's really just HTTP/2, with a protobuf payload. We can use the same URLs for both, making the switch-over easier.

The additional piece beyond that, is how/whether we can provide better support for gRPC features than we can for a purely call-and-response based HTTP/1.1 system. ie - providing streamed responses, and long-lived blocking execution of RPC calls (where we currently require polling).

**Comment by eddyashton** [comment]
Dumping some more thoughts and code here so I don't lose them.

The step from HTTP/1.1 to HTTP/2 is actually done already too - for demo purposes we can start a CCF service with `--http2`, and then use HTTP/2 clients. (When ALPN/merged endpoints from #3342 are done, we won't even need to manually turn this on - it will always be supported).

So the open question is "can we/should we build a shim to get a gRPC client talking JSON rather than protobuf?". I've built a hacky little thing that handles the serialization layer cheaply*, but it looks like we'd still need an interceptor to deal with the headers (because if CCF thinks it's speaking JSON, it inserts a "content-type: application/json" header in the response, and then the gRPC client blows up: `E0825 09:32:26.304294001   35396 hpack_parser.cc:1234]       Error parsing metadata: error=invalid value key=content-type value=application/json`).

In summary, I think we should just bite the bullet and go for full gRPC + protobuf-body support for the demo, as there's not that much work left to do here. I think we'll still be missing an async-response API, but that's fine for now.

<details>
  <summary>* Here's a bit of Python hackery that will dynamically generate a thing that looks like a `FooStub` from a `Foo` service definition, but using Python serialization:</summary>

```
from google.protobuf.json_format import MessageToJson, Parse

# Generates a class that looks like a Stub, but uses JSON serialization
def json_stub_class(module, service_name):
    service_descriptor = module.DESCRIPTOR.services_by_name[service_name]
    name = f"{service_name}Stub_JSON"

    def to_bytes(fn):
        def to(*args, **kwargs):
            ret = fn(*args, **kwargs)
            return ret.encode()

        return to

    def constructor(self, channel):
        for method, method_descriptor in service_descriptor.methods_by_name.items():
            outbound_arity = "stream" if method_descriptor.client_streaming else "unary"
            inbound_arity = "stream" if method_descriptor.server_streaming else "unary"
            uri = f"/{service_descriptor.full_name}/{method_descriptor.name}"
            channel_fn = getattr(channel, f"{outbound_arity}_{inbound_arity}")
            setattr(
                self,
                method,
                channel_fn(
                    uri,
                    request_serializer=to_bytes(MessageToJson),
                    response_deserializer=to_bytes(Parse),
                ),
            )
        pass

    methods = {"__init__": constructor}
    t = type(name, (object,), methods)
    return t


"""
Example use:

import grpc
import json_stub_class
import kv_pb2

creds = grpc.ssl_channel_credentials(open(path_to_service_cert, 'rb').read())
channel = grpc.secure_channel("127.0.0.1:8000", creds)
KVStub_JSON = json_stub_class.json_stub_class(kv_pb2, "KV")
kv_stub = KVStub_JSON(channel)
get_req = kv_pb2.KVKey(table="hello", key=b'1234')
kv_stub.Get(get_req)


TODO: Should probably just make the stub, rather than the class?
Though it's cool to know you can do the latter!
"""
```
</details>

**Comment by jumaffre** [comment]
I'll repurpose this issue to add support for gRPC in the external executor app. Tasks identified so far:

- [ ] Include and build libprotobuf
- [ ] Add `protoc` to ansible playbook
- [ ] Create `Get` and `Put` endpoints
- [ ] Add gRPC wrappers
